File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys

class Solution:
    def __init__(self):
        self.input = sys.stdin.readline
        self.output = sys.stdout.write
    
    def determination(self,lines,n,mid):
        res = 0
        for x in lines:
            res += x//mid
        return res >= n

    def answer(self):
        k, n = map(int, self.input().split())
        lines = []

        for _ in range(k):
            lines.append(int(self.input()))

        left, right, ans = 0, 2500000000, 0

        while left <= right:
            mid = (left + right) // 2

            if self.determination(lines, n, mid):
                ans = mid
                left = mid + 1
                continue

            right = mid - 1
        self.output(f'{ans}\n')

s = Solution()
s.answer()

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;
import java.util.*;

public class Main {

    private static boolean calFunc(int[] nums, long mid, int target ){
        long temp_res = 0;
        for(int num : nums){
            temp_res += num / mid; 

            if(temp_res >= target) return true;
        }
        return false;
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] temp = br.readLine().split(" ");        
        int k = Integer.parseInt(temp[0]);
        int n = Integer.parseInt(temp[1]);

        int[] nums = new int[k];

        for(int i = 0; i < k ; i++) nums[i] = Integer.parseInt(br.readLine());

        long left = 1;
        long right = Integer.MAX_VALUE;
        long mid = 0;
        long answer = 0;

        while (left <= right) {
            mid = (left + right) / 2;

            if(calFunc(nums, mid, n)){
                answer = mid;
                left = mid + 1;
                continue;
            }
            right = mid - 1;
        }

        System.out.println(answer);
        br.close();
    }
}